### ISSUE TITLE
Pushing input Events into actionscript for handling

### ISSUE THREAD

**Issue by Odie** [artumino/PapyrusVR#10]
Hi there!

I want to start a discussion about maybe pushing the raw controller input events to SkyUI for handling. For a better description of what I'm looking to accomplish, please see https://github.com/Odie/skyui-vr/issues/9. I wasn't sure if it is appropriate to go on at length about things not specifically relevant to PapyrusVR here.

In any case, it appears pushing data directly to scaleform/actionscript from c++ is not very difficult. The scaleform API is exposed via SKSE. In particular, ScaleformMovie::Invoke() can be used to call actionscript functions. Relevant Scaleform Documentation can be found [here](https://help.autodesk.com/view/SCLFRM/ENU/?guid=__scaleform_help_game_communication_c_actionscript_html).

**Comment by artumino** [comment]
Ok, I had a quick look at the Scaleform implementation in SKSE.
Yes, the Invoke is definitely what we need to send data back to UI components.

I also had a look at how to expose native functions to register for these events and everything we need is in the _Hooks_Scaleform.cpp_ file from SKSE. In the _InstallHooks_ function SKSE seems to allow plugins registered through _RegisterScaleformPlugin_ to register their custom native functions.

So each menu should be able to:

1. Register for events through our native functions by passing the callback function name that we'll Invoke with the event data.
2. Call native functions to get the state of SteamVR devices (pose, axis values, button states etc.)

**Comment by Odie** [comment]
This all sounds pretty reasonable.

I think the next part of the puzzle is being able to temporarily stop the game from handling VR button inputs while the menu is active. I suppose this might be doable from within SKSE somewhere. Are you familiar at all with where I might look or what I might try to get this to happen?

**Comment by artumino** [comment]
I'm not familiar with this too but I found something that might be of interest (even though it might cause lots of problems and might not even work).

From what I gather Skyrim handles VR keys as normal controllers through DirectInput (this assumption might be wrong). If the above assumption is correct we might be able to disable unwanted keypresses from the SKSE DirectInput hooks with:

`void	SetKeyDisableState(UInt32 keycode, bool bDisable, UInt32 mask);`

And then re-enable them once the menu is closed. 

Apart from the fact the it might not even work, I don't really like this solution cause It'll disable all type of inputs for every controller (potentially affecting players that play seated with only a headset and a normal controller).

With this hack-y solution we could send to the UI the user input device (Vive Wands, Oculus Touch, Generic Motion Controller or a common Controller) and let the UI create a blacklist of keys to disable. Then on menu close we would have to restore the state of the affected keys.

**Comment by artumino** [comment]
I also want to add that I don't know if thumbpad/trackpad interactions are handled as key presses in a direction or differently so disabling keys might not stop events coming from analogue inputs.

**Comment by Odie** [comment]
Another thing to try investigating is the MenuControls class, also exposed by SKSE.

It looks like a singleton with the following interesting members:
```
	MenuEventHandler* clickHandler;	// 48
	MenuEventHandler* directionHandler;	// 50
	MenuEventHandler* consoleOpenHandler;	// 58
	MenuEventHandler* quickSaveLoadHandler;	// 60
	MenuEventHandler* menuOpenHandler;	// 68
	MenuEventHandler* favoritesHandler;	// 70
	MenuEventHandler* screenshotHandler;	// 78
```

Maybe it's possible to replace the handler with something that can optionally disable all handling when the menu is active.

I have to say that it's not clear what any of these things do. Some experimentation will be required.

**Comment by lfrazer** [comment]
Hey in terms of input blocking for OpenVR I think I can do that by changing the values the game gets from PollNextEvent OpenVR lib function.  I will make a flag to enable/disable this behaviour.

**Comment by Odie** [comment]
That’d be great! At the experimental stage, I think I’d just go and drop in code until I found something that works. If anything comes of it at all, I’ll share the hacked up code so we can figure out what sort of api is needed/reasonable.

**Comment by Odie** [comment]
Quick update.

I spent a bit of time playing around with VRInputPlugin. Yes, it's possible to shut off the game's access to the button states quick easily via `FakeVRSystem::GetControllerState`. This bit of test code was used: https://github.com/Odie/SKSE-VRInputPlugin/commit/2213d53c880e0cf2db532b385eb612efa566b62b.

This uses the left touch pad clicks as a toggle to optionally stop the game from gaining access to new controller button states. In game, controller positions continued to update as expected. It was also verified that menus can be made to stop responding to all controller inputs with this mechanism.

It also appears that GetControllerState was being called at twice the framerate.

**Comment by lfrazer** [comment]
Ah yeah there are multiple ways the game could get input so I guess you found it, nice work.  I will try to implement something more official when I have time, hopefully this week.

**Comment by Odie** [comment]
Yeah, I'm glad that's the only place that needs to be blocked. Anyway, I'll try to push forward a bit more. There's a few more things to wire up.

**Comment by Odie** [comment]
Things are moving forward, though at a slow pace. Some work is being done here: https://github.com/Odie/SKSE-VRInputPlugin/tree/scaleform-api

This exposes functions to be called from scaleform. There are 3 functions exposed at the moment, meant to 1) toggle game's access to controller states 2) register/unregistered where to send new VR input events to.

Things should be done before too long. I'm leaving a note here so @lfrazer doesn't accidentally spend time doing something similar.

**Comment by Odie** [comment]
Hi all,

I got through implementing the initial version, which can be found here: https://github.com/Odie/SKSE-VRInputPlugin/tree/scaleform-api

Specifically, the following has been verified:
- Actionscript code can subscribe/unsubscribe for ControllerState updates.
- Actionscript code can shutoff ControllerState fetches from the game

Now that controller data can be pushed to actionscript, I can go work on connecting them to the existing handler code. Thanks everyone for making this happen!

BTW, please do let me know which repo to open the pull request against. =)

**Comment by lfrazer** [comment]
Nice work!  You can do pull request to my repo would be great to get those features added.
